### PR TITLE
Replace ad-hoc token verification logic by token.Verify

### LINKED ISSUE #559: Replace existing ad-hoc token creation and verification by a utils.CreateToken / utils.VerifyToken function
As one more step to finish #159 , we want to replace existing token creation logic in the codebase by one standard `utils.CreateToken` function to improve consistency.

Looking at `utils.CheckAnyAuth`, I think a couple of issues need to be addressed before the replacement:
* Rename the function to `utils.VerifyToken`
* Potentially rename the file name from `server_auth` to `token_validation`
* Besides the boolean return value, add a new return value to reflect the error or failure reason (if any) in the chain of checking
  * I'm thinking we can have a struct like 
  ```go
    type TokenValidationError struct {
        Type ValidationErrorType // string enum type, value = ValidationError or InvalidToken
        Message string // the failure reason/error message
    }
  ```
For `ValidationErrorType`, `ValidationError` is any error happened when validating the token, and  `InvalidToken` is due to incorrect/invalid token. We might want to distinguish this so that when we send the token validation failed http response back, we can give a clearer error message or even error code (400/500/401) 

* You might want to wait for #374 to be merged in first to as that PR changed the way `utils.CreateToken` works with scopes. #374 should be merged pretty soon.

### PR DESCRIPTION
**Pull request by haoming29**
Closes #559 and #572 

This PR is built upon #844. Open a draft PR as a placeholder for now. This should be ready once #844 is merged in.

### CODE REVIEW

**Comment by haoming29** [comment]
@jhiemstrawisc This is a follow-up issue to #844 so finish the refactoring of token verification logic, so I figured you might have the best context to review it.

File: token/token_verify.go [hunk 1]
```diff
@@ -267,53 +195,47 @@ func CheckAnyAuth(ctx *gin.Context, authOption AuthOption) bool {
...
 			} else {
-				log.Debug("Token validation succeeded with server issuer")
-				return exists
+				return http.StatusOK, true, nil
 			}
 		default:
-			log.Info("Authentication failed. Invalid/unsupported token issuer")
-			return false
+			log.Error("Invalid/unsupported token issuer")
+			return http.StatusInternalServerError, false, errors.New("Cannot verify token due to bad server configuration. Invalid/unsupported token issuer")
```
Comment (jhiemstrawisc) [inline token/token_verify.go:234]: If it's a bad token issuer, should this be a 403?

**Review by jhiemstrawisc — CHANGES_REQUESTED** [review]

File: token/token_verify.go [hunk 2]
```diff
@@ -70,82 +67,16 @@ const (
...
 }
 
-// [Deprecated] This function is expected to be removed very soon, after
-// https://github.com/PelicanPlatform/pelican/issues/559 is implemented
-//
-// Return director's public JWK for token verification. This function can be called
-// on any server (director/origin/registry) as long as the Federation_DirectorUrl is set
-//
-// The director's metadata discovery endpoint and JWKS endpoint are cached
-func LoadDirectorPublicKey() (jwk.Key, error) {
```
[the remark this answers is not in the document]
Comment (haoming29) [reply token/token_verify.go:90]: I don't think so. With other changes in the PR, this function is no longer being used, and this is a internal-facing helper function

File: token/token_verify.go (hunk 1 above)
Comment (haoming29) [reply token/token_verify.go:234]: The token issuer was set by us when evoking the verify function. Although a very rare situation, this is saying that somebody is not picking the issuer from the defined constants, so it's an internal error

File: token/token_verify.go [hunk 3]
```diff
@@ -267,53 +195,47 @@ func CheckAnyAuth(ctx *gin.Context, authOption AuthOption) bool {
...
+			log.Error("Invalid/unsupported token source")
+			return http.StatusInternalServerError, false, errors.New("Cannot verify token due to bad server configuration. Invalid/unsupported token source")
 		}
 	}
 
 	if token == "" {
-		log.Info("Authentication failed. No token is present from the list of potential token positions")
-		return false
+		log.Debug("Unauthorized. No token is present from the list of potential token positions")
+		return http.StatusUnauthorized, false, errors.New("Authentication is required but no token is present.")
```
[the remark this answers is not in the document]
Comment (haoming29) [reply token/token_verify.go:212]: I thought StatusUnauthorized is like you don't have the authorization token while StatusForbidden is that you _have_ the token but your token doesn't have correct permission. Maybe I'm wrong

File: origin_ui/origin_api.go [hunk 4]
```diff
@@ -121,28 +86,36 @@ func LaunchPeriodicDirectorTimeout(ctx context.Context, egrp *errgroup.Group) {
...
-		ctx.JSON(400, gin.H{"error": "Invalid director test response"})
+		ctx.JSON(http.StatusBadRequest, gin.H{"error": "Invalid director test response"})
 		return
 	}
 	// We will let the timer go timeout if director didn't send a valid json request
 	notifyNewDirectorResponse(ctx)
 	if dt.Status == "ok" {
 		metrics.SetComponentHealthStatus(metrics.OriginCache_Director, metrics.StatusOK, fmt.Sprintf("Director timestamp: %v", dt.Timestamp))
-		ctx.JSON(200, gin.H{"msg": "Success"})
+		ctx.JSON(http.StatusOK, gin.H{"msg": "Success"})
```
[the remark this answers is not in the document]
Comment (haoming29) [reply origin_ui/origin_api.go:112]: Agree. Brian did a first step in standardizing the response structure: https://github.com/PelicanPlatform/pelican/blob/4d653b942caa65f4590052041df5dec73685d343/broker/server_apis.go#L45. 

IIRC, we mostly use `{"msg":"Success"` or `{"message":"Success"}` in the code base, but it definitely worths a ticket to get them standardized.

File: director/redirect_test.go [hunk 5]
```diff
@@ -717,8 +717,8 @@ func TestDiscoverOriginCache(t *testing.T) {
 		w := httptest.NewRecorder()
 		r.ServeHTTP(w, req)
 
-		assert.Equal(t, 401, w.Code)
-		assert.Equal(t, `{"error":"Invalid token for accessing director's sevice discovery"}`, w.Body.String())
+		assert.Equal(t, 403, w.Code)
+		assert.Equal(t, `{"error":"Cannot verify token: Cannot verify token with server issuer:  Issuer is not server itself: https://wrong-issuer.org\n"}`, w.Body.String())
```
[the remark this answers is not in the document]
Comment (haoming29) [reply director/redirect_test.go:721]: I had trouble coming up with a clear description here so I admit I give up thinking. I think a better version can be, "token issuer does not match the local issuer on the current server"

File: token/token_verify.go (hunk 1 above)
Comment (jhiemstrawisc) [reply token/token_verify.go:234]: Cool, thanks for the explanation.

File: token/token_verify.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jhiemstrawisc) [reply token/token_verify.go:212]: Hmm I've seen that in a few places that try to help distinguish between a 401/403, too. I agree it's confusing My general concern with returning `401` is that the [RFC](https://www.rfc-editor.org/rfc/rfc9110#status.401) says we MUST also include the `WWW-Authenticate` with a challenge that provides a method for providing authentication credentials. I don't think we do that anywhere, however, and I don't understand it well enough to construct such a header.

Meanwhile the `403` specification contains this line:
```
If authentication credentials were provided in the request...
```

This leads me to believe that returning `403` in the absence of credentials is okay.

[the remark this answers is not in the document]
Comment (haoming29) [reply token/token_verify.go:212]: Makes sense to me.

**Comment by haoming29** [comment]
With the latest commit, I did a bunch of rename to make things clearer, `issuerCheck` is now called `localIssuerCheck` and the issuer constants are renamed to `FederationIssuer` and `LocalIssuer`

**Review by jhiemstrawisc — APPROVED** [review]
LGTM

**Pull request merged**